### DISCUSSION TITLE
IntelliFlo VF protocol error & rpm reporting

### DISCUSSION THREAD

**Discussion by 99to** [tagyoureit/nodejs-poolController#1004] [Nixie]
I am considering converting an IntelliTouch system to standalone Nixie. My system includes two Pentair P/N 011012 IntelliFlo VF pumps (which I understand are not that common and have been discontinued for some time) and having gathered from a few previous discussions that they are accordingly less tested with njsPC, I decided to setup a test instance to confirm pump operation including a single pump connected to a 485 gateway and everything else isolated.

Using this test setup, I discovered what I believe is one minor protocol issue and one nice-to-have missing item.

1) When the pump is commanded off (either by turning off the associated virtual circuit or entering service mode), njsPC appears to be repeatedly trying to send an 'action 5/Set[feature]' command with an empty payload which results in 'Error sending setPumpFeature' errors in the console log and an unexpected payload of '1' in the return message.

2) Although the VF pump is not designed to accept speed commands over the RS485 bus, it does report the current speed in the status request responses. With IntelliTouch/ScreenLogic, this actual speed is reported along with power and the commanded (vs. actual) flow. In my tests, while the rpm does appear to be updating as expected in bytes 5 and 6 of the pump status payload (as documented here: https://github.com/tagyoureit/nodejs-poolController/wiki/Pumps), it does not appear on the dashPanel pump status field nor does the poolState.json reflect these values (rpm is fixed at 0). Is it possible to configure/modify this so the reported rpm is available for integrations and dashPanel display, similarly to the VSF pump (although without the ability to set an rpm value of course)?

Here is a replay archive from the Nixie test setup described above containing a test sequence with a speed ramp up and down, then pump off and on commands and ending with auto->service and service->auto transitions:

[Nixie-VF_pump-replay-20240719.zip](https://github.com/user-attachments/files/16317209/Nixie-VF_pump-replay-20240719.zip)

In addition, here is a packet log that I captured with the IntelliTouch OCP communicating with the same pump following the same general test sequence above. I have reviewed/compared both and while there are some timing and order differences (timing looks better to me on Nixie by the way) the only item of concern was the empty payload commands described in 1 above where in the case of IntelliTouch, all Set[feature] commands have a single byte payload of '6' and a corresponding acknowledgement from the pump, irrespective of whether or not the pump is running:

[IntelliTouch-VF_pump-packetLog-20240719.zip](https://github.com/user-attachments/files/16317214/IntelliTouch-VF_pump-packetLog-20240719.zip)

The Message Manager tool was sure appreciated- it made this analysis quick and pleasant, many thanks for the thought and time that went into implementing that!

Finally, for the benefit of any other VF pump users, I did have a single instance of a nearly instantaneous over current alarm on the pump display panel on one transition from auto to service mode. Based on the discussion here: https://github.com/tagyoureit/nodejs-poolController/discussions/756 and never having seen an over current in almost 10 years on either pump, I suspect this is somehow related to the timing or content of the RS485 comms. The only clue I have is that sometimes, but not repeatably, when entering service mode the pump control panel transitions from the automation lockout screen to the 'manual' mode rather than the 'feature 1' mode. I've also observed it switch to 'feature 1' with the pump re-starting right away and 600 mins remaining showing, whereas most of the time with Nixie, and every observed time with the IntelliTouch, it transitions to 'feature 1' with 0 mins remaining and the pump stopped. It seems plausible this might be related to the Set[feature] issue above but unfortunately I was unable to reproduce this over current error, so am not suggesting pursuing it any further, just wanted to document what I did find here in case it comes up again for anyone else.

**Comment by rstrouse** [comment]
I will have to go back and see if I can figure out where we got this empty payload from.  I believe it was because the not releasing control back to its control panel.  We did this for some reason specifically for VF pumps.  Perhaps this was a misunderstanding because I do not have a VF pump.  We did pin down the overcurrent in the previous post to the priming attributes on the pump.

I removed the empty payload from the VF output so give this a shot.  I also posted the rpm back to the poolState.json but I have examples of that data being incorrect on the payload.  Perhaps this is more of a firmware version difference though.

**Reply by rstrouse** [reply]
Oh pull njsPC to see the changes.  Unfortunately, I do not have the equipment to even see what the differences are other than your outputs from replay.

**Reply by 99to** [reply]
That change ended up causing the pump to cycle on and off in a way that sounded similar to the previous discussion in the over current thread however I did find what I believe is the solution.

After looking again at the capture I took yesterday from IntelliTouch communicating with a VF pump I suspect the command order matters as well, in particular, where the commands setting flow and feature should be sent prior to turning the pump on. Secondly, and a clue to why the empty payload that was in place before mostly worked (although was throwing errors), the OCP wasn't sending any feature command when off (i.e. it never sends any feature payload other than 6, but sometimes doesn't send action 5 at all).

Now that I see from your modification this code is specific to only the VF pump, it makes sense to me to just match what the IntelliTouch is doing which, using the syntax in the comment above your code modification, is:

When the pump is off, send 4[255], 6[4]
When the pump is on, send 4[255], 1[flow], 5[6], 6[10]

I have tested this by changing the order and conditions in that class as described above and with those changes it now appears to be working as expected, no console errors and no observed pump errors or unexpected starts/stops.

The rpm is now updating in the poolState.json and looks to be correct (although it does not appear in dashPanel).

In the pumps section of the poolState.json the status is still showing 'error12' (just as in the replay sent yesterday). Is this used elsewhere or important and if so, would you like me to help try to chase down the issue there?

**Reply by rstrouse** [reply]
Pull njsPC and dashPanel.  I made the sequence changes identified above and pushed a change to dashPanel to show the RPM.

**Reply by 99to** [reply]
Thanks. With these sequence changes, the pump is now responding as expected in all tested modes and the rpm is displaying on the dashPanel and looks correct (although fyi, I believe there is a typo in the associated code edit).

This leaves the only remaining issue I'm aware of the 'error12' status mentioned earlier.

I have investigated this a little and determined:
- it looks like the status appearing in poolState.json is being pulled from the pump status reply message payload bytes 11 and 12, which is consistent with the table here: https://github.com/tagyoureit/nodejs-poolController/wiki/Pumps#pump-status-message
- however, at least on this VF, bytes 11 and 12 do not contain fault status but instead the hours and minutes set for feature 1. I was able to confirm this by changing (on the pump control panel) from the Feature 1 default of 600 minutes to 598 and verified the corresponding change from [10,0] to [9,58].
- when interrupting power to the pump (which still leaves the pump controller active for ~15 seconds or so), byte 10 changed from 0 to 128 which suggests to me this is very likely at least one of the fault status bytes and the faults are represented as a bitmap, with bit 8-power loss fault and the normal running or stopped condition represented by all bits cleared. I didn't note any non-zero values in bytes 8 or 9 in any of my testing, so it is plausible that 9 and 10 are a 16 bit fault register, but that is unverified.
- with respect to normal status, the following was consistent in my testing today:
	- if byte 0 = 4, then status = Stopped
	- if byte 0 = 10 & byte 2 = 1, then status = Priming
	- if byte 0 = 10 & byte 2 = 4, then status = System Priming
	- if byte 0 = 10 & byte 2 = 3, then status = Primed (present only for ~2 seconds)
	- if byte 0 = 10 & byte 2 = 2, then status = Running

I'm not sure how much any of this matters to the rest of njsPC, but wanted to at least document what I'd found. At a minimum, it seems like a good idea to stop pulling status from the feature time fields for this pump.

To document the changes to the pump command sequences as well as these status/error messages, here is a replay with the pump programmed to 1 minute System Priming Time and a Feature 1 timeout of 9 hours, 58 minutes, with the end of the sequence showing the response to interrupting pump power: [Nixie-VF_pump-replay-20240721.zip](https://github.com/user-attachments/files/16326422/Nixie-VF_pump-replay-20240721.zip)

**Reply by rstrouse** [reply]
The weird thing with this is that our original research states that when the drive state (byte 2) is 0 then there is a fault and the status bytes 11 and 12 contain a fault status.  What we were unaware of is that the time schedule is being reflected into those bytes.  Perhaps this little tidbit was missed in the coding of the state message.

Bytes 9 & 10 have always been a mystery.  I wouldn't be surprised if the 128 is actually a power outage alarm and we were just off by two bytes in the sequence.  I absolutely believe that we got a bit turned around on the pump reporting so I will make a change here.

Byte 13 & 14 are the total runtime for the pump but it will roll over.  This kind of makes sense if one were trying to predict the off time for the pump using the minutes calculation from the schedule set on the pump.  This is because there is no time clock on the pump only a timing circuit.  One could record the pump time when it is turned on then determine when the pump will go off based on the schedule.  However, we do not use the pre-programmed schedules in the pump so bytes 11 & 12 have no real meaning for us.

**Reply by rstrouse** [reply]
This is what we currently have documented for the pump status message including the changes expressed above.

Index | Value | Description
-- | -- | --
0 | Command | I suspect that this is   actually a bitmask, but for now we can identify 4 as being off and 10 as   running.  Somewhere in here I expect to find a freeze bit.<br/>1=??????<br/> 2=Power on<br/>3=??????<br/>4=Running<br/>5=??????<br/>6=??????   <br/>7=??????<br/>8=??????
1 | Mode | This is programmed mode that   is currently selected on the pump. 1-8 for most pumps <br/>0 = Filter<br/>1 = Manual<br/>2 = Backwash<br/>6 = Feature 1?<br/>9 = Ext Program 1<br/>10 = Ext Program 2<br/>11 = Ext Program 3<br/>12 = Ext Program 4<br/>13 = Ext Program 5<br/>14 = Ext Program 6<br/>15 = Ext Program 7<br/>16 = Ext Program 8<br/>255 = No mode
2 | Drive State | The values below are from an old documentation of the   protocol.<br/>0 = Fault Mode (When in   this state the fault bits on the status will be valid)<br/>1 = Priming<br/>2 = Running<br/>4 = System Priming
3-4 | Energy in Watts | (byte(3) * 256) + byte(4)   = Number of watts.
5-6 | Speed | (byte(5) * 256) + byte(6)   = Pump rpm.
7 | Flow | The current flow in the   units stored on the pump.  This will   either be gpm or lpm.
8 | Ppc | Not sure what the ppc value   is but it has to do with the pump pressure curve.  This is used to determine what the current   back pressure on the pump is to calculate the filter values.
9-10 | Status | 16 bits of error and   status codes for the pump with the most severe being the highest order bit.<br/>0=Ok<br/>   1=Ok - Some pumps report 1 for ok as well<br/>2=Filter Warning<br/>3=Overcurrent<br/>4=Priming Error<br/>5=System Blocked   <br/>6=General Alarm<br/>7=Overtemp Condition<br/>8=Power Outage<br/>9=Overcurrent 2 Condition<br/>10=Overvoltage Condition<br/>11=???????<br/>12=???????<br/>13=???????<br/>14=???????<br/>15=???????<br/>16=Communication Failure
11-12 | Program time |  The number of minutes set for the program on feature 1
13-14 | Time | The current time set on   the pump.
  |   |  




</body>

</html>

**Reply by 99to** [reply]
Confirming/clarifying a few items just for the documentation record (agreed none of these timing related fields should matter in the context of njsPC):

- Bytes 11 and 12 are definitely the hours/minutes set for feature 1 although note this is not a running timer- it reflects the preset runtime programmed in the pump for feature 1 but does not decrement while the pump is running via RS485 commands.

- Bytes 13 and 14, at least in the VF pump I am testing with, actually are tied to a real time clock. The current time can be set with the pump control panel (and appears on the manual mode display in the top right corner). The IntelliTouch also sets the time on the pump and I've now confirmed that is done using the pump protocol, action 3 (which I haven't seen documented) and two payload bytes, hours and minutes (using 24h time). I have been able to replicate this by sending that sequence when running Nixie which is immediately reflected in the pump status messages and then on the manual display once the control panel is released. Note that this action 3 sequence appears to only work if the pump is stopped, otherwise what I assume is an error packet (action 255) is returned by the pump.

I can also confirm that the byte 2 drive state does **not** reflect a 0 when the power fault occurs. In fact, it would appear that the drive state does not update (probably isn't valid) when the pump is stopped (i.e. byte 0 = 4). For example, if the pump was priming and then interrupted, byte 2 will remain set to 1 for the duration it is stopped, likewise for system priming. So in the chart (at least as far as this VF pump with its version of firmware, no idea if the protocol might have changed between firmware versions) under byte 2, I would suggest removing the reference to 0 and the fault mode and adding in that 3 represents 'Primed' (which is momentary only, not latching, seems to match the way it appears on the LCD when running manually) along with a comment that the drive state is only valid when byte 0 = 10.

If you do change any of the status reporting logic I would be happy to confirm that it is working on this pump. I really appreciate your time working through this with me and on the project in general- many thanks.

**Reply by rstrouse** [reply]
For posterity can I get a replay where the action 3 is being sent to the pump to include the response.

**Reply by 99to** [reply]
This replay reflects using action 3 to change the pump time from the ~current time to 12:34, waiting for a minute rollover then changing back to the current time: [Nixie-VF_pump-clock_set-replay-20240724.zip](https://github.com/user-attachments/files/16366302/Nixie-VF_pump-clock_set-replay-20240724.zip)
